### ISSUE TITLE
Enable setting a global value for A4 to affect midi - frequency conversion

### ISSUE THREAD

**Issue by gesellkammer** [csound/csound#694]
Currently all opcodes converting between midi note and frequency (like cpsmidinn) are hardwired to A4=440 Hz. It would be nice to be able to set a global value for A4 to match the tuning of live instruments.

**Comment by jpffitch** [comment]
Not trivial as all over and there is a table built in aops.  Do you want this at init time or dynamic?
Or compiled in?

./Engine/fgens.c:                           + (double) lpd.detune * 0.01) / 12.0) * 440.0;
./Engine/memfiles.c:                                         + (int) lpd.detune) / 1200.0) * 440.0;
./OOps/aops.c:    *p->r = pow(FL(2.0), (*p->a - FL(69.0)) / FL(12.0)) * FL(440.0);
./Opcodes/dssi4cs/src/ladspa.h:   should be used. This will be 440 unless the host uses an unusual
./Opcodes/fm4op.c:    p->baseFreq = FL(440.0);
./Opcodes/fm4op.c:      {FL(3270.0), FL(0.440), -FL(17.0)}},
./Opcodes/fm4op.c:      {FL(3270.0), FL(0.440), -FL(17.0)}},
./Opcodes/modal4.c:/*     m->baseFreq = 440.0; */
./Opcodes/pinker.c:#define PINK_BIAS   440.0f
./Opcodes/sfont.c:      globals->pitches[j] = (MYFLT) (440.0 * pow(2.0, (double)(j- 69)/12.0));
./frontends/csladspa/ladspa.h:   should be used. This will be 440 unless the host uses an unusual

**Comment by vlazzarini** [comment]
I was thinking of an OPARMS field

**Comment by vlazzarini** [comment]
or a system constant like Odbfs

**Comment by vlazzarini** [comment]
replace all occurrences of 440
for it and we're done

**Comment by jpffitch** [comment]
No; if 440 changes ten the aops table neds regeneration.  If this is at 
load time then it is OK; if changable then need an opcode to regenerate. 
So which?

**Comment by gogins** [comment]
I removed this table some time ago for cpsmidinn, in the interests of
greater precision. The code now is:

int cpsmidinn(CSOUND *csound, EVAL *p)
{
    IGN(csound);
    *p->r = pow(FL(2.0), (*p->a - FL(69.0)) / FL(12.0)) * FL(440.0);
    return OK;
}

This opcode could easily be changed to take an optional i_a4 parameter that
would default to 440.

Regards,
Mike


-----------------------------------------------------
Michael Gogins
Irreducible Productions
http://michaelgogins.tumblr.com
Michael dot Gogins at gmail dot com

**Comment by vlazzarini** [comment]
Not an opcode; a constant just like sr etc. Not changeable after Csound is constructed, only after reset.

**Comment by jpffitch** [comment]
ok code writtenl just how to set te value not done, and code in pinker.c I do not follow

**Comment by jpffitch** [comment]
Should tis be a command-line option or an r-variable
--A4=440
or
  a
A4 = 440

**Comment by gesellkammer** [comment]
I personally would like it to be a variable like nchnls or 0dbfs.

**Comment by gogins** [comment]
That works for me.

Regards,
Mike

**Comment by kunstmusik** [comment]
I would like to vote against a global variable.  The engine has a enough special handling code for ksmps, sr, etc. and I would rather not see further cases introduced.  Wouldn't two opcodes for setting and getting the A4 value be sufficient?

**Comment by gesellkammer** [comment]
On Thu, Sep 1, 2016 at 3:56 AM, Steven Yi <notifications@github.com> wrote:

> I would like to vote against a global variable. The engine has a enough
> special handling code for ksmps, sr, etc. and I would rather not see
> further cases introduced. Wouldn't two opcodes for setting and getting the
> A4 value be sufficient?
>
> —
>

An opcode is just as good (something like "setA4 442"). A getter is not
really necessary, since one can just convert 69 to cps to get that value.

**Comment by vlazzarini** [comment]
I think a variable is a cleaner way to do it and it is transparent to the user.

**Comment by kunstmusik** [comment]
I have some troubles with this, at least in regards to the implementation.  

1. Why is this stored in OPARMS when other reserved values are held in the CSOUND struct?  
2. Do we use this implementation as a precedent for the next user request for global configuration values?   

The internal developmental changes seem at odds with the precedents of existing code.  If I had to write up about the engine (which I am working on now), I would find this hard to explain.  For example:

"Most global values are held within the CSOUND struct, but not all. (Check OPARMS, which is for user configured values, generally from the commandline or via API calls, which may also hold global values referenced by opcodes.) Those exposed to the user are put into the global VAR_POOL. There is special case handling in the parser for reserved values like sr, ksmps, etc. to allow only a single assignment.  To add new global values, modify the code base in x, y, and z locations instead of using...."

This isn't to say I'm completely against having this as a user-exposed variable, but I do see problems.  The suggestion of opcodes was meant to hide these kinds of issues and perhaps allow time to consider the bigger picture.  If we move ahead with the code as-is, then I would vote that we mark this code as something to consider for re-implementation and spend some time to design how global values like A4 could be handled in a more generic, extensible way. (Ideally, 3rd-party opcode library developers would also be able to define their own global configuration values that only their opcodes may know about, and that might include exposing setting the value via command-line, registering in the global var pool, etc.)

**Comment by vlazzarini** [comment]
I see A4 on a par with 0dbfs and in my opinion should be implemented in a similar way.

**Comment by kunstmusik** [comment]
I understand the rationale behind making it a variable but I think we need to consider the costs too.  For example, we have setksmps as an opcode, but it's used as a variable too.  We have this ingrained now into the system.  We also have things like UDOs handling arguments by value, where native opcodes pass by reference.  These kinds of things are difficult to change now.  All I am arguing at this point is that an opcode in this case, although less easy to use for users, would have been safer in terms of internal changes. We could have exposed a variable later based upon some kind of designed out system for handling this kind of problem more generally.  

However, the change is already done. I've raised my concerns and I will leave it here now.

**Comment by vlazzarini** [comment]
I don't think an opcode is suitable in this case. Setksmps is an opcode to set ksmps, so the variable is ksmps, which is held locally. If we decide to handle system constants such as sr,kr,ksmps,0dbfs,nchnls_i,nchnls differently then we should handle A4 differently. It's of enough importance to be at level (not a mere configuration variable).

**Comment by jpffitch** [comment]
Now seems to be working and I cleaned up some of the implementation

Still to do is remove the cpsoctfrac table an compute directly but I am closing the issue as it is fixed
